codec/src/main/java/io/netty/handler/codec/compression/ZstdDecoder.java

```diff
@@ -0,0 +1,149 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import com.github.luben.zstd.ZstdInputStreamNoFinalizer;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+
+/**
+ * Decompresses a compressed block {@link ByteBuf} using the Zstandard algorithm.
+ * See <a href="https://facebook.github.io/zstd">Zstandard</a>.
+ */
+public final class ZstdDecoder extends ByteToMessageDecoder {
+    private final MutableByteBufInputStream inputStream = new MutableByteBufInputStream();
+    private ZstdInputStreamNoFinalizer zstdIs;
+
+    private State currentState = State.DECOMPRESS_DATA;
+
+    static {
+        try {
+            Zstd.ensureAvailability();
+        } catch (Throwable throwable) {
+            throw new ExceptionInInitializerError(throwable);
+        }
+    }
+
+    /**
+     * Current state of stream.
+     */
+    private enum State {
+        DECOMPRESS_DATA,
+        CORRUPTED
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        try {
+            if (currentState == State.CORRUPTED) {
+                in.skipBytes(in.readableBytes());
+                return;
+            }
+            final int compressedLength = in.readableBytes();
+
+            inputStream.current = in;
+
+            ByteBuf outBuffer = null;
+            try {
+                int w;
+                do {
+                    if (outBuffer == null) {
+                        // Let's start with the compressedLength * 2 as often we will not have everything
+                        // we need in the in buffer and don't want to reserve too much memory.
+                        outBuffer = ctx.alloc().heapBuffer(compressedLength * 2);
+                    }
+                    do {
+                        w = outBuffer.writeBytes(zstdIs, outBuffer.writableBytes());
+                    } while (w != -1 && outBuffer.isWritable());
+                    if (outBuffer.isReadable()) {
+                        out.add(outBuffer);
+                        outBuffer = null;
+                    }
+                } while (w != -1);
+            } finally {
+                if (outBuffer != null) {
+                    outBuffer.release();
+                }
+            }
+        } catch (Exception e) {
+            currentState = State.CORRUPTED;
+            throw new DecompressionException(e);
+        } finally {
+            inputStream.current = null;
+        }
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        super.handlerAdded(ctx);
+        zstdIs = new ZstdInputStreamNoFinalizer(inputStream);
```

Comment:
the current usage will allocate internal byte[] for each instantiation which will contribute to memory pressure, but there is a option to use a BufferPool: https://github.com/luben/zstd-jni/blob/20014d19a91a1b04e4e198ac1cb85b5acb1c1142/src/main/java/com/github/luben/zstd/ZstdInputStreamNoFinalizer.java#L62

RecyclingBufferPool.INSTANCE is a stock implementation, but it may be possible to create an implementation which is backed by the netty's PooledAllocator if that is desired.

Reply:
here is skeletal BufferPool implementation which is backed by Netty's allocator

```java
public class ZstdBufferPool implements BufferPool {
    // the constructor seems to use only a single buffer
    private final List<ByteBuf> byteBufs = new ArrayList<>(1); 
    @Override
    public ByteBuffer get(int capacity) {
      ByteBuf byteBuf = alloc.heapBuffer(capacity);
      byteBufs.add(byteBuf);
      return byteBuf.nioBuffer();
    }
    @Override
    public void release(ByteBuffer buffer) {
      // pass, instead release these buffers after we close stream
    }
    // call this after the stream is closed
    public void releaseAll(){ 
      for(ByteBuf byteBuf: byteBufs){
        bufBufs.release();
      }
    }
}
```

Reply:
actually nevermind, the implementation does not appear to support ByteBuffers which are slices of larger arrays, so I guess RecyclingBufferPool.INSTANCE is the only option.

https://github.com/luben/zstd-jni/blob/20014d19a91a1b04e4e198ac1cb85b5acb1c1142/src/main/java/com/github/luben/zstd/Zstd.java#L1538

https://github.com/luben/zstd-jni/blob/20014d19a91a1b04e4e198ac1cb85b5acb1c1142/src/main/java/com/github/luben/zstd/ZstdInputStreamNoFinalizer.java#L164

Reply:
I think we should pull it in as it is ... we can do a follow for improvements later